Animation script, which makes an animation between two prompts

machineinthegod opened the pull request:
This is somewhat different than just img2img loopback methods (like deforum).
It gets more coherent animation and tries to spread the changes between as many frames as possible.

How it works: 
The entire animation is made from a single noise, with different weighting of the two prompts for each frame. 
That's not enough for coherency because there will still be many jumps if you just vary the weighting between the two prompts.
So instead the frames are generated like a tree:
The first sampling steps are done on a small number of frames, because they correspond to bigger changes, and the later sampling steps are done on a bigger number of frames, because they correspond to smaller changes.
Somewhat like a tree, close frames share common ancestor which makes their details similar.
It also makes generation of many frames much faster than if you had to do the whole sampling process for each frame.
Some variations in the weighting between the prompts produce few changes, and others produce many. To deal with that, I look at the squared difference between different frames, and frames with higher differences will have more frames between them in later iterations.
There's also an option in img2img to generate a video starting directly from the image given, so that you could chain an animation.


Changes required outside the script:
I needed to modify the sampler to be able to partially sample. All of this works only on KDiffusion samplers. 
I added video display to display the output of this script, which is hidden unless the result is returned.
I added an option to use TotalTQDM to display the progress bar (because the current way it's implemented doesn't fit)
All the animation is seen as a single batch by process_images, and it required some changes so that it won't crash on VRAM memory while decoding it, or from not finding the prompt.

In windows you might need to download OpenH264 Video Codec DLL (openh264-1.8.0-win64.dll) from cisco because cv2 video encoding library requires it and it doesn't come built-in. Frame generation works regardless, and you can also generate AVI without it but gradio doesn't support displaying it, only h264.

Commit:
Created animation script, which generates somewhat smoother animation between two prompts than loopback img2img. Required changes to UI to support video output, changes to progress bar to display progress properly, and some hooks around sampling to be able to share initial sampling results between different animation frames.

Commit:
fixed img2img issues in animation

Commit:
Merge branch 'master' of github.com:machineinthegod/stable-diffusion-webui

ClashSAN commented:
Can you make a separate repository so we can add this to the custom scripts page?
Like this: https://github.com/yownas/seed_travel

machineinthegod commented:
@ClashSAN  Like this https://github.com/machineinthegod/stable-diffusion-webui ?

It does require few modifications to the main repo to accommodate, so just the script alone isn't enough.
(All the changes to the main repo are explained in the pull request).

machineinthegod commented:
https://user-images.githubusercontent.com/113234863/197420107-0e8d447d-20bf-44f8-ab46-980bce580158.mp4

machineinthegod commented:
That's 64 frames between 
Mount Everest, HDR color 4k by national geographic, Nikon camera, cloudy, high resolution, award winning photography
and
(Volcanic eruption), smoke, ashes, fireballs, lava, HDR color 4k by national geographic, raytracing, detailed fire

You can see there are no jumps / random features popping into existence from nowhere, there are barely any changes between individual frames. (this was 30 FPS, all frames made by the script)

machineinthegod commented:
https://user-images.githubusercontent.com/113234863/197420565-1269555e-3fda-49cc-b337-7b9a62be374a.mp4

This is at 2 FPS so you can appreciate how little it changes between individual frames.

ClashSAN commented:
It's very nice!

>https://github.com/machineinthegod/stable-diffusion-webui

Not that way, as no user could leave issues or feature requests. You can also give a readme description for instructions on how to use. You can put your module folder with your changed files along with the animator script.

Your script seems to involve many modifications to the original repo files. any chance you can reduce such a requirement with time ?

LiJT commented:
Can't wait this merge into main repo

papuSpartan commented:
maybe the image viewer will become the *viewer*

machineinthegod commented:
The core of the main repo changes is required, because I need to be able to partially sample (not all steps at once), and the main repo is missing my changes to KDiffusion. 

Rest of the changes are ensuring I won't crash the main code and integrating with the UI in a sensible way (displaying video, same filename format for videos, progress bar which works, stuff like that).

I can make a pull request with only the main repo changes if that helps.

Commit:
merged with master

Commit:
changed the default value of img2img option

AUTOMATIC1111 commented:
The script should definitely be an extension.

Adding a hidden video element to the UI is fair, getting just the filename with save image is fair, non-batched decode_first_stage is fair.

`if state.interrupted or state.skipped:` should not be there, sampling already solves that. `seed_i`, `prompt_i` don't seem right to me, if you want multiple pictures you specify multiple seeds and prompts.

Sampling has `stop_at`; if that's not enough for you, just hook the samplers yourself in your script and return them to how they were afterwards. i don't understand changes in sampler so I really hope that it's possible to not introduce them. Why do we not add noise in some cases? Can't you get same effect by just passing 0 as noise?

I don't understand why the change to TotalTQDM is needed.

Commit:
removed lines which I accidentially reintroduced during merge

machineinthegod commented:
`if state.interrupted or state.skipped:` was accidentally reintroduced during a merge (it was there before me and deleted while I was working), I fixed it.

I do sampling for the animation in a special way: I don't sample everything at once - I partially sample, and then resume sampling on the same intermediate samples with different conditioning each time. So the first steps (sigmas - because they corresponding to sigmas in the array passed to the internal model sampling routine) are done on very few frames, while the later sampling steps are done on a much higher number of frames. 
`stop_at` isn't enough, because there's no way to resume sampling. 
`sigma_end_index` is somewhat like `stop_at`, but without an exception, it's somewhat more readable because it's always coming with sigma_start_index. I suppose I could skip it and rewrite if you still think it's necessary after you understand the rest.
`sigma_start_index` is like resuming after stop_at. As the code was written before, there was no way to resume after stopping sampling. I don't add the noise a second time because that's the way sampling would've proceeded if it had gone uninterrupted. Also notice that I don't multiply by sigmas[0] if we're resuming in the middle - this is important.

I also need the values of sigma outside the sampler. Since the code computing sigmas was duplicated, I moved it to the function `get_sigmas`. The code acts identical to before if `sigma_start_index==0`, but if it isn't, it acts as if you resumed sampling in the middle. I think it's better when it's deduplicated anyway.

Now the rest of the changes with prompt_i and seed_i is because I do the sampling of multiple frames all at once, all sharing the same single prompt and seed, I will return more samples than batch_size (which is 1 usually). I can try to hack away from this change maybe by lying about batch size, if necessary. 

I just made an option to use TotalTQDM for the gradio UI progress bar, which also solves a problem I had. TotalTQDM works great because you can set an arbitrary total steps at updateTotal, while the logic at `check_progress_call` implicitly assumes that job_count * steps is the total number of steps and works badly in my case (Progress bar will jump around inaccurately without it). I bet there are other scripts out there with similar problems that just didn't bother fixing it.

cvar66 commented:
So is this going to be made into an extension?
